tls/s2n_connection_serialize.c

```diff
@@ -0,0 +1,158 @@
+
+    /* Best effort check for pending input or output data.
+     * This method should not be called until the application has stopped sending and receiving.
+     * Saving partial read or partial write state would complicate this problem.
+     */
+    POSIX_ENSURE(s2n_stuffer_data_available(&conn->header_in) == 0, S2N_ERR_INVALID_STATE);
+    POSIX_ENSURE(s2n_stuffer_data_available(&conn->in) == 0, S2N_ERR_INVALID_STATE);
+    POSIX_ENSURE(s2n_stuffer_data_available(&conn->out) == 0, S2N_ERR_INVALID_STATE);
+
+    uint32_t context_length = 0;
+    POSIX_GUARD(s2n_connection_serialization_length(conn, &context_length));
+    POSIX_ENSURE(buffer_length >= context_length, S2N_ERR_INSUFFICIENT_MEM_SIZE);
+
+    struct s2n_blob context_blob = { 0 };
+    POSIX_GUARD(s2n_blob_init(&context_blob, buffer, buffer_length));
+    struct s2n_stuffer output = { 0 };
+    POSIX_GUARD(s2n_stuffer_init(&output, &context_blob));
+
+    POSIX_GUARD(s2n_stuffer_write_uint64(&output, S2N_SERIALIZED_CONN_V1));
+
+    POSIX_GUARD(s2n_stuffer_write_uint8(&output, conn->actual_protocol_version / 10));
+    POSIX_GUARD(s2n_stuffer_write_uint8(&output, conn->actual_protocol_version % 10));
+    POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secure->cipher_suite->iana_value, S2N_TLS_CIPHER_SUITE_LEN));
+
+    POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secure->client_sequence_number, S2N_TLS_SEQUENCE_NUM_LEN));
+    POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secure->server_sequence_number, S2N_TLS_SEQUENCE_NUM_LEN));
+
+    POSIX_GUARD(s2n_stuffer_write_uint16(&output, conn->max_outgoing_fragment_length));
+
+    if (conn->actual_protocol_version >= S2N_TLS13) {
+        uint8_t secret_size = 0;
+        POSIX_GUARD(s2n_hmac_digest_size(conn->secure->cipher_suite->prf_alg, &secret_size));
+
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secrets.version.tls13.client_app_secret, secret_size));
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secrets.version.tls13.server_app_secret, secret_size));
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secrets.version.tls13.resumption_master_secret, secret_size));
+    } else {
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->secrets.version.tls12.master_secret, S2N_TLS_SECRET_LEN));
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->handshake_params.client_random, S2N_TLS_RANDOM_DATA_LEN));
+        POSIX_GUARD(s2n_stuffer_write_bytes(&output, conn->handshake_params.server_random, S2N_TLS_RANDOM_DATA_LEN));
+    }
+
+    return S2N_SUCCESS;
+}
+
+int s2n_connection_deserialize(struct s2n_connection *conn, uint8_t *buffer, uint32_t buffer_length)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(conn->secure);
+
+    struct s2n_blob context_blob = { 0 };
+    POSIX_GUARD(s2n_blob_init(&context_blob, buffer, buffer_length));
+    struct s2n_stuffer input = { 0 };
+    POSIX_GUARD(s2n_stuffer_init_written(&input, &context_blob));
+
```

Comment:
Should we check that there's at least as much data as we expect? Otherwise the caller will get a stuffer error instead of a more helpful "bad input" error. Alternatively, you could have a wrapper method translate some errors.

Reply:
🤔 Probably the only check we could do here is this:
POSIX_ENSURE_GTE(buffer_length, S2N_TLS13_FIXED_STATE_SIZE +  smallest_secret * 3);
Like ensure that the buffer is larger than the smallest state possible. Otherwise we don't really know how much data to expect until you read the protocol version and cipher suite.

Reply:
Maybe the wrapper to translate errors is the best options then